services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/kafka/KafkaBasedMappingAndDelegatingCommandHandler.java

```diff
@@ -12,47 +12,86 @@
  *******************************************************************************/
 package org.eclipse.hono.commandrouter.impl.kafka;
 
+import java.net.HttpURLConnection;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
 import java.util.Objects;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.function.Supplier;
 
+import org.apache.kafka.common.TopicPartition;
+import org.eclipse.hono.adapter.client.command.CommandContext;
 import org.eclipse.hono.adapter.client.command.kafka.KafkaBasedCommand;
 import org.eclipse.hono.adapter.client.command.kafka.KafkaBasedCommandContext;
 import org.eclipse.hono.adapter.client.command.kafka.KafkaBasedInternalCommandSender;
 import org.eclipse.hono.adapter.client.registry.TenantClient;
+import org.eclipse.hono.client.ServerErrorException;
 import org.eclipse.hono.client.impl.CommandConsumer;
 import org.eclipse.hono.client.kafka.tracing.KafkaTracingHelper;
 import org.eclipse.hono.commandrouter.CommandTargetMapper;
 import org.eclipse.hono.commandrouter.impl.AbstractMappingAndDelegatingCommandHandler;
+import org.eclipse.hono.tracing.TracingHelper;
 
 import io.opentracing.Span;
 import io.opentracing.SpanContext;
 import io.opentracing.Tracer;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
 import io.vertx.core.buffer.Buffer;
 import io.vertx.kafka.client.consumer.KafkaConsumerRecord;
 
 /**
  * Handler for commands received at the tenant-specific topic.
  */
 public class KafkaBasedMappingAndDelegatingCommandHandler extends AbstractMappingAndDelegatingCommandHandler {
+
+    /**
+     * Timeout applied to determining target gateway and adapter instance for a command, <i>if</i> that
+     * command is blocking the delegation of a newer command for which these values have already
+     * been determined. An elapsed timeout in that scenario means that the processing of the
+     * corresponding command is cancelled, so that the next-in-line command can be delegated to its
+     * target adapter.
+     */
+    static final Duration BLOCKING_COMMAND_QUEUE_ITEM_TIMEOUT = Duration.ofSeconds(30);
```

Comment:
The javadoc above describes a safeguard feature I've added here.
I wanted to handle a case where the `getTargetGatewayAndAdapterInstance` operation (which includes device registry and data grid requests) is stuck and blocking the processing of later commands for which `getTargetGatewayAndAdapterInstance` has already returned a value.
I've therefore implemented a timeout mechanism to remove/cancel such command records so that the later commands can get sent.

However, on second thought, I think this is probably not really needed here. The data grid request duration could be limited via HotRod-client timeout configuration values and a scenario where one request is stuck for that long and later ones have already succeeded is probably not very realistic.

I'll therefore remove this feature here, making the code a lot simpler.